### PR TITLE
Implement real middleware and allow custom global error handler

### PR DESCRIPTION
**Pull request by zRosenthal**
* middleware is no able to be executed in a specific order
* add ability to supply custom error handler
* all errors thrown by mindless are now of type `MindlessError`

### CODE REVIEW

File: src/routing/router.ts [hunk 1]
```diff
@@ -18,17 +16,13 @@ export class Router<
...
+  protected static getRouteMetaData(route: Route<Middleware, Controller>): { [key: string]: any } {
     /**
      * controller and middleware are constructors
      * there should be no need for them
      */
     const isUsefulKey = (key: string) =>
-      typeof (route as any)[key] !== 'undefined' &&
-      key !== 'controller' &&
-      key !== 'middleware'
+      typeof (route as any)[key] !== 'undefined' && key !== 'controller' && key !== 'middleware'
```
Comment (shametim) [inline src/routing/router.ts]: check out 7.1.2 here http://exploringjs.com/es6/ch_symbols.html

Comment (zRosenthal) [reply src/routing/router.ts]: not a bad idea but outside the scope of this change

Comment (zRosenthal) [reply src/routing/router.ts]: feel free to open an issue or pr or both 😉

File: src/app/dispatcher.ts [hunk 2]
```diff
@@ -33,29 +41,22 @@ export class Dispatcher {
...
-        'Error Message': e.message,
-        'Mindless Message':
-          'Unable to resolve requested controller or method make sure your _routes are configured properly'
-      }
-      return new Response(500, body)
+      throw new MindlessError(
+        `Failed to resolve controller from container. Error Message: ${e.message}`
+      )
     }
+    const args = params.map(getArgToInject)
```
Comment (moraneva) [inline src/app/dispatcher.ts]: put this inside `try` so that `subjectController` can be const?

File: src/error/default-error-handler.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { Request } from '../request'
+import { Response } from '../response'
+import { CustomErrorHandler } from './custom-error-handler'
+
+export const DefaultErrorHandler: CustomErrorHandler = (e: Error, request: Request): Response => {
+  console.error(e)
+  return new Response(500, {
+    message:
+      'an error occurred, using default error handler (see docs to supply your own). Please check your logs'
```
Comment (moraneva) [inline src/error/default-error-handler.ts]: `An`

File: src/app/dispatcher.ts (hunk 2 above)
Comment (zRosenthal) [reply src/app/dispatcher.ts]: ya but i want the error from mapping params to bubble up so then i would have to 
```
try {
// ...
const args = params.map(getArgsToInject)
// ...
} catch (e) {
  if (e instance of MindlessError) {
    throw e
  }
  throw new MindlessError('')
}

**Pull request merged**